EIPS/eip-7212.md

```diff
@@ -0,0 +1,180 @@
+
+# b: elliptic curve coefficient
+0x5ac635d8aa3a93e7b3ebbd55769886bc651d06b0cc53b0f63bce3c3e27d2604b
+
+# G: base point of the subgroup
+(0x6b17d1f2e12c4247f8bce6e563a440f277037d812deb33a0f4a13945d898c296,
+ 0x4fe342e2fe1a7f9b8ee7eb4a7c0f9e162bce33576b315ececbb6406837bf51f5)
+
+# n: order of the subgroup
+0xffffffff00000000ffffffffffffffffbce6faada7179e84f3b9cac2fc632551
+
+# h: cofactor of the subgroup
+0x1
+
+```
+
+### Elliptic Curve Signature Verification Steps
+
+The signature verifying algorithm takes the signed message hash, the signature components provided by the “secp256r1” curve algorithm, and the public key derived from the signer private key. The verification can be done with the following steps:
+
+```
+# h (message hash)
+# pubKey = (public key of the signer private key)
+
+# Calculate the modular inverse of the signature proof:
+s1 = s^(−1)(modn)
+
+# Recover the random point used during the signing:
+R' = (h * s1) * G + (r * s1) * pubKey
+
+# Take from R' its x-coordinate:
+r' = R'.x
+
+# Calculate the signature validation result by comparing whether:
+r' == r
+
+```
+
+### Required Checks in the Verification
+
+Some requirements have to be checked to understand if the given signature components are valid:
+
+- Verify that both `r` - `s` values are greater than 0 and less than the curve order.
+- Verify that s is equal to or less than half of the order of the subgroup to prevent signature malleability.
+- Verify that the point formed by (`x`, `y`) values are on the curve.
+
+### Precompiled Contract Specification
+
+The `P256VERIFY` precompiled contract is proposed with the following input and outputs:
+
+- **Input data:** 160 bytes of data including:
+    - 32 bytes of the signed data `hash`
+    - 32 bytes of the `r` component of the signature
+    - 32 bytes of the `s` component of the signature
+    - 32 bytes of the `x` coordinate of the public key
+    - 32 bytes of the `y` coordinate of the public key
+- **Output data:** 32 bytes of result data and error
+    - If the signature verification process succeeds, it returns 1 in 32 bytes format, if not it returns 0 in 32 bytes format for the result with a null error.
+    - If the given `x` and `y` coordinates are not valid, it returns null for the result with the given error.
+
+### Precompiled Contract Gas Usage
+
+The use of signature verification cost by `P256VERIFY` is `3450` gas. Following reasons and calculations are provided in the [Rationale](#rationale) and [Test Cases](#test-cases) sections.
+
+## Rationale
+
+The "secp256r1" elliptic curve signatures consists of `v`, `r`, `s` components. Even if recovering the public key on the curve is possible, most of the applications are not generating `v` component of the signature and it causes an uncertainty of the result values. However, the signatures can be verified with only `r` - `s` values. In order to provide an exact and more compatible method, verification is preferred over recovery to propose in a precompiled.
```

Comment:
It would be nice to clarify exactly what we gain from requiring `x`/`y` as inputs.  It's not clear to me from the explanation here.

And requiring them as inputs requires substantial added call data.

Reply:
If I understand correctly,

- `v` is a sign bit
- the P256 precompile is intended to support hardware signatures
- most of the existing hardware signature mechanisms (Secure Enclave, Webauthn, etc) return only the (r,s) signature

easy to recover v -- you try the two possible values, and see which one returns your public key.

so the comparison is something like

### `v, r, s` style

Pros:
- **Matches ecrecover.**
- **Less calldata.** On the surface, 1 word = 32 bytes less, but actual savings should be a bit more, since the `v` parameter is mostly 0 bytes and therefore cheaper.

Cons:
- **Users / libraries have to recover `v`.** This is a bit awkward. Instead of just calling (hardware signing module), it has to call (hardware signing module) > (attempt verification with + v) > (attempt verification with - v). I am not a cryptographer obv, I am not sure if there's a better way.

### `v, r, x, y` style

Pros:
- **Better matches the hardware interface.** Users/libraries don't have to recover `v`, they just pass a pubkey (x,y) directly from the hardware enclave.

Reply:
FYI https://eips.ethereum.org/EIPS/eip-2098 for  might be relevant in the same spirit where user / library recovers `v` for ECDSA. 

This is provided to you as a reference and you can gauge the preference by usage / adoption for designing the precompiles

Reply:
Thanks @dcposch for the explanation of the logic.

Yes, the mechanisms using the curve are returning only `r` and `s`, so the verification is possible with inputting public key coordinates `x` and `y`.

We need `v` value to recover public key without the `x` and `y` coordinates. The `v` value can be found by above-mentioned methods, but let me explain my design choice:

- While we are making recovery with the `ecrecover`, we can reach the public address of the EOA accounts, so it can be directly used in the smart contracts. Unlikely, recovering the `secp256r1` public key does not match any default stored types, and we still need to store the account public key.
- Having to find the `v` value in the implementation part of the signature creates complexity on the application side. So, I didn't want to bring this complexity for the applications. Still, I would love to reassess and edit the EIP to implement recovery after discussion.